inc/azure_macro_utils/macro_utils.h

```diff
@@ -78,73 +78,101 @@ MU_IF(X, "true", "false") => "true"
 #define MU_EAT_EMPTY_ARG(arg_count, arg) MU_IF(MU_ISEMPTY(arg),,arg) MU_IF(MU_ISEMPTY(arg),MU_EXPAND_TO_NOTHING,MU_IFCOMMALOGIC)(MU_DEC(arg_count))
 #define MU_EAT_EMPTY_ARGS(...) MU_FOR_EACH_1_COUNTED(MU_EAT_EMPTY_ARG, __VA_ARGS__)
 
+#define MU_EMPTY()
+#define MACRO_UTILS_DELAY(id) id MU_EMPTY MU_LPAREN )
+
+//
+//
+// MU_DEFINE_ENUM
+//
+//
+
+// This uber enum macro can handle enumeration-constants being constant expressions.
+// Given the following enum:
+//
+//
+// #define TEST_UBER_ENUM_VALUES \
+//     test_uber_a, \
+//     test_uber_b = 5
+// 
+// MU_DEFINE_ENUM(TEST_UBER_ENUM, TEST_UBER_ENUM_VALUES)
+//
+// The ToString expands to something like the below given an enum
+//
+// const char* TEST_UBER_ENUM_ToString(TEST_UBER_ENUM value)
+// {
+//     static int my_test_uber_a;
+//     static int my_test_uber_b = 5;
+//     static int enum_last_value = -1;
+//     enum_last_value = my_test_uber_a ? (my_test_uber_a) : (enum_last_value + 1);
+//     if (enum_last_value == value)
+//     {
+//         return "test_uber_a";
+//     }
+//     enum_last_value = (my_test_uber_b = 5) ? (my_test_uber_b = 5) : (enum_last_value + 1);
+//     if (enum_last_value == value)
+//     {
+//         return "test_uber_a";
+//     }
+//     return "NULL";
+// }
```

Comment:
why is this comment necessary in the .h file, it can go in the documentation very easily. None of the other macros have such an extensive documentation.

Reply:
The reason I placed it here is for the code reader/reviewer's sanity. I can remove it if you insist :-).

---
In reply to: [328346940](https://github.com/Azure/azure-macro-utils-c/pull/30#discussion_r328346940) [](ancestors = 328346940)